File: hello/src/leetcode/regex/Q010.java
package leetcode.regex;

/** 
 * 10. Regular Expression Matching
 * 
 * https://leetcode.com/problems/regular-expression-matching/solution/
 * <pre>
	Given an input string (s) and a pattern (p), implement regular expression matching with support for '.' and '*'.

	'.' Matches any single character.
	'*' Matches zero or more of the preceding element.
	The matching should cover the entire input string (not partial).

	Note:

	s could be empty and contains only lowercase letters a-z.
	p could be empty and contains only lowercase letters a-z, and characters like . or *.

	Example 1:
	Input: s = "aa" p = "a"
	Output: false
	Explanation: "a" does not match the entire string "aa".

	Example 2:
	Input: s = "aa" p = "a*"
	Output: true
	Explanation: '*' means zero or more of the preceding element, 'a'. Therefore, by repeating 'a' once, it becomes "aa".

	Example 3:
	Input: s = "ab" p = ".*"
	Output: true
	Explanation: ".*" means "zero or more (*) of any character (.)".

	Example 4:
	Input: s = "aab" p = "c*a*b"
	Output: true
	Explanation: c can be repeated 0 times, a can be repeated 1 time. Therefore, it matches "aab".

	Example 5:
	Input: s = "mississippi" p = "mis*is*p*."
	Output: false
	</pre>
 * @author odys-z@github.com
 *
 */
public class Q010 {

	/**
	 * Runtime: 69 ms, faster than 10.85% of Java online submissions for Regular Expression Matching.
	 * Memory Usage: 39.7 MB, less than 37.37% of Java online submissions for Regular Expression Matching.
	 * @param s
	 * @param p
	 * @return
	 */
	public boolean isMatch1(String s, String p) {
		int slen = s == null ? 0 : s.length();
		if (p == null || p.length() == 0)
			return s == null || slen == 0;

		char p0 = p.charAt(0);
		boolean m0 = s != null && slen > 0
					&& (p0 == s.charAt(0) || p0 == '.');
		
		if (p.length() > 1 && p.charAt(1) == '*')
			return isMatch1(s, p.substring(2))
					|| (m0 && slen > 0 && isMatch1(s.substring(1), p));
		else return m0 && slen > 0 && isMatch1(s.substring(1), p.substring(1));
	}

	/**Remove substring orperations.
	 * 
	 * Runtime: 24 ms, faster than 33.31% of Java online submissions for Regular Expression Matching.
	 * Memory Usage: 38 MB, less than 47.47% of Java online submissions for Regular Expression Matching.
	 * @param s
	 * @param p
	 * @return
	 */
	public boolean isMatch2(String s, String p) {
		int slen = s == null ? 0 : s.length();
		int plen = p == null ? 0 : p.length();
		if (plen == 0)
			return slen == 0;
		
		return match(s, 0, slen, p, 0, plen);
	}
	
	private boolean match(String s, int sx, int slen, String p, int px, int plen) {
		if (px >= plen) return sx >= slen;
		if (sx >= slen)
			// now, px < plen, only ".*.*..." is ok
			return px + 2 <= plen && p.charAt(px + 1) == '*' && match(s, sx, slen, p, px + 2, plen);
		char p0 = p.charAt(px);

		boolean m0 = slen > 0 && (p0 == s.charAt(sx) || p0 == '.');
		
		if (px + 1 < plen && p.charAt(px + 1) == '*')
			return match(s, sx, slen, p, px + 2, plen)
					|| (m0 && match(s, sx+1, slen, p, px, plen));
		else return
			m0 && (sx+1 < slen ? match(s, sx+1, slen, p, px+1, plen) : 
								match("", 0, 0, p, px+1, plen));
	}
	
	private int slen = 0;
	private int plen = 0;
	private String s;
	private String p;
	private Boolean[][] ij;
	/**DP
	 * Runtime: 1 ms, faster than 100.00% of Java online submissions for Regular Expression Matching.
	 * Memory Usage: 39.3 MB, less than 43.43% of Java online submissions for Regular Expression Matching.
	 * @param s
	 * @param p
	 * @return
	 */
	public boolean isMatch(String s, String p) {
		slen = s == null ? 0 : s.length();
		plen = p == null ? 0 : p.length();
		if (plen == 0)
			return slen == 0;
		
		this.s = s;
		this.p = p;
		this.ij = new Boolean[this.slen+1][this.plen+1];
		return matchDp(0, 0);
	}

	private boolean matchDp(int i, int j) {
		if (ij[i][j] == null) {
			if (j >= plen) return i >= slen;
			else if (i >= slen)
				return j+2 <= plen && p.charAt(j+1) == '*' && matchDp(i, j+2);
			else {
				char p0 = p.charAt(j);
				// m0
				boolean m0 = slen > 0 && (p0 == s.charAt(i) || p0 == '.');
				
				if (j+1 < plen && p.charAt(j+1) == '*')
					ij[i][j] = matchDp(i, j+2)
								|| m0 && matchDp(i+1, j);
				else ij[i][j] = m0 && matchDp(i+1, j+1);
			}
		}
		return ij[i][j];
	}
}

File: hellopy/test/leetcode/regex/q010.py
import unittest

'''
10. Regular Expression Matching
https://leetcode.com/problems/regular-expression-matching/

Runtime: 1548 ms, faster than 7.77% of Python3 online submissions for Regular Expression Matching.
Memory Usage: 13.8 MB, less than 5.55% of Python3 online submissions for Regular Expression Matching.
'''
class Solution2:
    def isMatch(self, s: str, p: str) -> bool:
        plen = len(p) if p is not None else 0
        slen = len(s) if s is not None else 0
        if p is None or plen == 0:
            return s is None or slen == 0

        m0 = slen > 0 and (p[0] == s[0] or p[0] == '.')

        if plen > 1 and p[1] == '*':
            return self.isMatch(s, p[2:]) or (m0 and self.isMatch(s[1:], p))
        else:
            return m0 and self.isMatch(s[1:], p[1:])

slen = 0
plen = 0
'''
Runtime: 1156 ms, faster than 26.37% of Python3 online submissions for Regular Expression Matching.
Memory Usage: 13.8 MB, less th
'''
class Solution3:
    def isMatch(self, s: str, p: str) -> bool:
        global plen, slen
        plen = len(p) if p is not None else 0
        slen = len(s) if s is not None else 0
        if plen == 0:
            return slen == 0

        return self.match(s, 0, p, 0)

    def match(self, s: str, sx: int, p: str, px: int):
        global plen, slen
        if px >= plen:
            return sx >= slen
        
        m0 = slen > sx and (p[px] == s[sx] or p[px] == '.')

        if px < plen-1 and p[px+1] == '*':
            return self.match(s, sx, p, px+2) or (m0 and self.match(s, sx+1, p, px))
        else:
            return m0 and self.match(s, sx+1, p, px+1)

ij = None;
'''
Runtime: 36 ms, faster than 94.87% of Python3 online submissions for Regular Expression Matching.
Memory Usage: 13.6 MB, less than 5.55% of Python3 online submissions for Regular Expression Matching.
'''
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        global plen, slen, ij
        plen = len(p) if p is not None else 0
        slen = len(s) if s is not None else 0
        if plen == 0:
            return slen == 0
        ij = [None] * (slen+1)
        return self.matchDp(s, 0, p, 0)

    def matchDp(self, s: str, sx: int, p: str, px: int):
        global plen, slen, ij
        if px >= plen:
            return sx >= slen
        
        if (ij[sx] is None):
            ij[sx] = [None] * (plen+1)
            
        if (ij[sx][px] is None):
            m0 = slen > sx and (p[px] == s[sx] or p[px] == '.')

            if px < plen-1 and p[px+1] == '*':
                ij[sx][px] = self.matchDp(s, sx, p, px+2) or (m0 and self.matchDp(s, sx+1, p, px))
            else:
                ij[sx][px] = m0 and self.matchDp(s, sx+1, p, px+1)
        return ij[sx][px]

class Test(unittest.TestCase):
    def test1(self):
        s = Solution()
        self.assertTrue(s.isMatch(None, None))
        self.assertTrue(s.isMatch("", None))
        self.assertTrue(s.isMatch("", ""))
        self.assertTrue(s.isMatch("", ".*"))
        self.assertTrue(s.isMatch(None, ".*"))
        self.assertTrue(s.isMatch('', '.*'))
        self.assertFalse(s.isMatch("", "a"))
        self.assertTrue(s.isMatch("", "a*"))
        self.assertTrue(s.isMatch("", "c*c*"))
        self.assertFalse(s.isMatch("a", ""))
        self.assertTrue(s.isMatch('a', '.'))
        self.assertTrue(s.isMatch("a", "a"))
        self.assertFalse(s.isMatch("aa", "a"))
        self.assertTrue(s.isMatch("ab", ".b"))
        self.assertTrue(s.isMatch("ab", ".*"))
        self.assertTrue(s.isMatch("aaa", ".*"))
        self.assertTrue(s.isMatch("aab", "c*a*b"))
        self.assertFalse(s.isMatch("mississippi", "mis*is*p*."))
        self.assertTrue(s.isMatch( "mississippi", "mis*is*ip*."))
        self.assertTrue(s.isMatch( "mississi", "mis*is*i*"))
        self.assertFalse(s.isMatch("mississi", "mis*is*i."))
        self.assertTrue(s.isMatch( "mississi", "mis*is*i.*"))
        self.assertTrue(s.isMatch( "mississi", "mis*is*i"))
        self.assertTrue(s.isMatch("{\"\":\"ississ123}", "{\"\":\"is*is*123}"))
        self.assertTrue(s.isMatch("{\"\":  123}", "{\"\": *123}"))
        self.assertTrue(s.isMatch("mississi", "...*is*i"))
        self.assertTrue(s.isMatch("mississi", ".*i.*i.*i.*"))
        self.assertTrue(s.isMatch("iii", ".*i.*i.*i.*"))
        self.assertTrue(s.isMatch("1i2ii4", ".*i.*i.*i.*"))
        self.assertTrue(s.isMatch("1i2i i", ".*i.*i.*i.*"))
        self.assertTrue(s.isMatch("i i iss", ".*i.*i.*i.*"))
        self.assertFalse(s.isMatch("i", ".*i.*i"))
